service/src/event/ordering_task.rs

```diff
@@ -175,302 +97,470 @@ impl OrderingTask {
-        if undelivered.is_empty() {
-            return Ok((0, 0));
-        }
-        let found = undelivered.len();
-        let mut new = 0;
-        for (key, data) in undelivered {
-            let event_cid = key.cid().ok_or_else(|| {
-                Error::new_invalid_arg(anyhow::anyhow!("EventID is missing a CID: {}", key))
-            })?;
-            let (insertable_body, maybe_prev) =
-                CeramicEventService::parse_event_carfile(event_cid, &data).await?;
-            if let Some(prev) = maybe_prev {
-                let event = DeliverableEvent::new(insertable_body.cid, prev, None);
-                if self.track_pending(event).is_some() {
-                    new += 1;
+        event_data: Vec<(EventId, Vec<u8>)>,
+    ) -> Result<usize> {
+        trace!(cnt=%event_data.len(), "Processing undelivered events batch");
+        let mut event_cnt = 0;
+        let mut discovered_inits = Vec::new();
+        for (_event_id, carfile) in event_data {
+            let (cid, parsed_event) =
+                unvalidated::Event::<Ipld>::decode_car(carfile.as_slice(), false)
+                    .await
+                    .map_err(Error::new_app)?;
+
+            let metadata = EventMetadata::from(parsed_event);
+
+            let (stream_cid, loaded) = match &metadata {
+                EventMetadata::Init => {
+                    discovered_inits.push(cid);
+                    continue;
                 }
-            } else {
-                // safe to ignore in tests, shows up because when we mark init events as undelivered even though they don't have a prev
-                info!(event_cid=%insertable_body.cid, "Found undelivered event with no prev while processing undelivered. Should not happen. Likely means events were dropped before.");
-                self.ready_events.push_back(insertable_body.cid);
-                new += 1; // we treat this as new since it might unlock something else but it's not actually going in our queue is it's a bit odd
-            }
-        }
-        trace!(%new, %found, "Adding undelivered events to pending set");
-        Ok((new, found))
-    }
+                EventMetadata::Data { stream_cid, prev }
+                | EventMetadata::Time { stream_cid, prev } => (
+                    stream_cid,
+                    StreamEvent::Undelivered(StreamEventMetadata::new(cid, *prev)),
+                ),
+            };
 
-    fn add_incoming_batch(&mut self, events: Vec<DeliverableEvent>) -> HashSet<InitCid> {
-        let mut updated_streams = HashSet::with_capacity(events.len());
-        for event in events {
-            if let Some(updated_stream) = self.track_pending(event) {
-                updated_streams.insert(updated_stream);
+            event_cnt += 1;
+            self.add_stream_event(*stream_cid, loaded);
+        }
+        // while undelivered init events should be unreachable, we can fix the state if it happens so we won't panic in release mode
```

Comment:
this seems reasonable to me. I think its okay to do a little more at startup to try to recover from an unexpected condition, but continue to bail if something impossible happens at runtime.